src/Functions/parseDateTime.cpp

```diff
@@ -1035,6 +1035,37 @@ namespace
                 return cur;
             }
 
+            static Pos mysqlMicrosecond(Pos cur, Pos end, const String & fragment, DateTime & /*date*/)
+            {
+                checkSpace(cur, end, 6, "mysqlMicrosecond requires size >= 6", fragment);
+
+                Pos start = cur;
+                auto check_is_number = [&](Pos pos)
+                {
+                    if (*pos < '0' || *pos > '9')
+                        throw Exception(
+                            ErrorCodes::CANNOT_PARSE_DATETIME,
+                            "Unable to parse fragment '{}' from '{}' because '{}'' is not a number ",
+                            fragment,
+                            std::string_view(start, end),
+                            *cur);
+                };
+
+                check_is_number(cur);
+                ++cur;
+                check_is_number(cur);
+                ++cur;
+                check_is_number(cur);
+                ++cur;
+                check_is_number(cur);
+                ++cur;
+                check_is_number(cur);
+                ++cur;
+                check_is_number(cur);
+                ++cur;
```

Comment:
for loop

Reply:
also what's up with `readNumber{2..4}`, why not extend it for 6?
And why can't it be a single function `template<size_t N>` with for loop inside?
I trust compiler should be able to do loop unwinding with constant loop size.

Reply:
for loop: done

`readNumber()` actually reads the number, whereas here, we want to read *over* the number (ignore it). But I'll add function `assertNumber()` which should make this more beautiful.

Reply:
why does `mysqlSecond` then read the number? what's the difference?